### ISSUE TITLE
Support multiple examples specified in Open API schema

### ISSUE THREAD

**Issue by Stranger6667** [kiwicom/schemathesis#589]
As mentioned in #582 we can generate multiple examples via the `examples` keyword:

```
parameters:
  - in: query
    name: limit
    schema:
      type: integer
      maximum: 50
    examples:       # Multiple examples
      zero:         # Distinct name
        value: 0    # Example value
        summary: A sample limit value  # Optional description
      max: # Distinct name
        value: 50   # Example value
        summary: A sample limit value
```

**Comment by Stranger6667** [comment]
And I am very happy to hear that you'd like to implement it :) Let me know if I can provide support on this

**Comment by barrett-schonefeld** [comment]
@Stranger6667 sounds great! I will start exploring the approaches you mentioned and will see your approach to GraphQL when you open a PR. Just to clarify, I plan to implement the version where we pair examples arbitrarily:

    request(limit=0, id=1), request(limit=50, id=2), request(limit=0, id=3)


Also, do we need to consider the possibility of property-level examples in request bodies as discussed in #467, or will #467 be implemented later?

**Comment by Stranger6667** [comment]
I think that we can start with limited scope for now and expand it later :) But it will be good if we'll arrange the code for this feature in the way so it will be easier to extend it later with #467

**Comment by barrett-schonefeld** [comment]
@Stranger6667 it looks like the `jsonschema` package does not support the `examples` keyword.

![Screen Shot 2020-06-05 at 8 13 35 AM](https://user-images.githubusercontent.com/59850348/83880799-56bb1e00-a705-11ea-880e-24085443b1fa.png)

I looked at the JSON Schema specification and found that it does have an `examples` keyword, but they treat the value of `examples` as an array. 


    "examples" : [
        "Anything",
        4035
    ]

I'm going to open an issue against `hypothesis-jsonschema`, and I hope they will consider supporting `examples without validating the value of `examples`.

**Comment by Stranger6667** [comment]
Hm, it doesn't seem related to the `jsonschema` package, since we have JSON Schema definitions for Open API copied from the official sources, and there are `examples` keywords in the relevant places. What schema are you validating?

I am also not sure how `hypothesis-jsonschema` is related to it since it generates valid data for the passed schema. Could you, please, provide more info?

**Comment by Stranger6667** [comment]
Based on [these definitions](https://github.com/kiwicom/schemathesis/blob/master/src/schemathesis/specs/openapi/definitions.py) we have the following

Open API 2:

`examples` is an object with an unlimited amount of properties

Only `examples` inside response:

```
"200": 
    description: Test
    examples: 
      foo: ...
```

Open API 3:

There are more places where it is possible to put `examples`.

1. inside `Components` which is probably not relevant
2. inside `MediaType`, basically in the response or parameter serialization
3. in `Parameter`

`Example` in Open API 3 has some more validation, but it is an object still:

```
        {
            "type": "object",
            "properties": {
                "summary": {"type": "string"},
                "description": {"type": "string"},
                "value": {},
                "externalValue": {"type": "string", "format": "uri-reference"},
            },
            "patternProperties": {"^x-": {}},
            "additionalProperties": False,
        }
```

What version of the spec you were working with?

**Comment by barrett-schonefeld** [comment]
You're right. I thought the `examples` were defined under `schema` the same way an `example` is defined. This is what I had that produces the exception above.

    components:
        schemas:
            SimpleSchema:
                type: object
                description: an object with a string field
                required:
                    - foo
                properties:
                    foo:
                        type: string
                examples:
                    example1:
                        value:
                            foo: this is the first example!
                    example2:
                        value:
                            foo: this is the other example!

I did not notice the difference when reviewing the OpenAPI spec, but I moved the `examples` "up" one level, and the schema was accepted.

**Comment by barrett-schonefeld** [comment]
@Stranger6667 I made a bit of progress on this issue, but I am stuck on trying to understand where the code should live and how to handle a `requestBody` with multiple `mediaTypes`.

My strategy is to use the `EndpointDefinition` to collect all of the `examples` from `parameters` and collect all of the `examples` from `requestBody` and then merge the examples to form a list of `static_parameter` dictionaries.

So far, I have written code for this in `_hypothesis.py`. This is code is incomplete. I am just showing the code to retrieve all of the `examples` from `parameters`:

    def get_static_params_from_examples(endpoint: Endpoint)-> Dict[str, Any]:
        endpoint_def = endpoint.definition.raw
        all_parameter_examples = []
        for param in endpoint_def.get("parameters", []):
            if "examples" in param:
                # add value to list of params with examples
                param_type = {"header": "headers", "cookie": "cookies", "path": "path_parameters", "query": "query"}.get(param["in"])
                assert param_type in PARAMETERS, f"param type: {param_type} not in {PARAMETERS}"
                all_parameter_examples.append({
                    "type": param_type,
                    "name": param["name"],
                    "examples": [example["value"] for example in param["examples"]],
                })
    ...

Questions:

1. Currently, this code is in `_hypothesis.py`. Do we need to implement this in `specs.openapi.schemas` to distinguish between OpenAPI 3 and Swagger 2 implementations? Should we implement this in Swagger 2 using `x-examples` because `examples` are only supported for responses? This is what I'm thinking:

    class Endpoint:
    ...
        def static_params_from_examples(self)-> Dict[str, Any]:
            return self.schema.static_params_from_examples(self)

2. How should we retrieve examples from `requestBody` with multiple `mediaTypes`? Should we collect `examples` from all `mediaTypes`? Or does Schemathesis only use one `mediaType` for testing (if, so how do we know which `mediaType` is being used)? 

        requestBody:
            content:
                application/json:
                    examples: ...
                    schema: ...
                multipart/form-data:
                    examples: ...
                    schema: ...
        ...

3. This question assumes we should collect `examples` from all `mediaTypes`. Should  `body` and `form_data`be mutually exclusive in the `static_parameters` object? As in, we should not have this:

    static_parameters = {
        ...
        "body": {
            ...
        },
        "form_data": {
            ...
        }

4. Is there a well-defined mapping Schemathesis that shows which `mediaTypes` map to the `body` parameter and which `mediaTypes` map to the `form_data` parameter?

**Comment by Stranger6667** [comment]
Hi @barrett-schonefeld 

I'll reply later today or tomorrow

**Comment by barrett-schonefeld** [comment]
@Stranger6667 hey, just want to check in. I plan to continue with this work tomorrow.

**Comment by Stranger6667** [comment]
Hi @barrett-schonefeld 

Sorry for not getting back to you.

1. Yep, I think we can put it to `specs/openapi/examples.py` and then connect it in the way you mentioned.
2. Currently, we don't handle multiple media types - only the first one defined in `content` or `application/json` in `openapi/serialization.py`. Probably we can leave it for later and support only the first media type - supporting multiple media types in all areas will require much more work.
3. I think we can have only one media type for now
4. Oh, no, I don't think we have it right now

Let me know if I can provide more info

**Comment by Stranger6667** [comment]
Implemented in #620
